src/ReinforcementLearningCore/src/policies/q_based_policies/learners/approximators/neural_network_approximator.jl

```diff
@@ -151,6 +153,160 @@ function (model::GaussianNetwork)(state, action)
+    da = size(μ,1)
+    L = trilcol(cholesky_vec,da)
+    z = Zygote.ignore() do
+        noise = randn(rng, eltype(μ), da, action_samples, batch_size)
+        model.normalizer.(Flux.stack(map(.+, eachslice(μ,dims=3), eachslice(L, dims=3) .* eachslice(noise,dims=3)),3)) 
+    end
+    logp_π = mvnormlogpdf(μ, L, z)
+    return z, logp_π
+end
+
+function (model::CovGaussianNetwork)(state::AbstractArray, args...; kwargs...)
+    model(Random.GLOBAL_RNG, state, args...; kwargs...)
+end
+
+"""
+    (model::CovGaussianNetwork)(state, action)
+Return the logpdf of the model sampling `action` when in `state`. 
+State must be a 3D tensor with dimensions (state_size x 1 x batch_size).
+Multiple actions may be taken per state, `action` must have dimensions (action_size x action_samples_per_state x batch_size)
+Returns a 3D tensor with dimensions (1 x action_samples_per_state x batch_size)
+"""
+function (model::CovGaussianNetwork)(state::AbstractArray, action::AbstractArray)
+    da = size(action,1)
+    x = model.pre(state)
+    μ, cholesky_vec = model.μ(x), model.Σ(x)
+    L = trilcol(cholesky_vec,da)
+    logp_π = mvnormlogpdf(μ, L, action)
+    return logp_π
+end
+
+"""
+If given 2D matrices as input, will return a 2D matrix of logpdf. States and actions are paired column-wise, one action per state.
+"""
+function (model::CovGaussianNetwork)(state::AbstractMatrix, action::AbstractMatrix)
+    output = model(Flux.unsqueeze(state,2), Flux.unsqueeze(action,2))
+    return dropdims(output, dims = 2)
+end
+
+"""
+Transform a vector containing the non-zero elements of a lower triangular da x da matrix into that matrix.
+"""
+function trilcol(cholesky_vec,da)
+    batch_size = size(cholesky_vec, 3)
+    c2idx(i,j) = ((2da-j)*(j-1))÷2+i #return the position in cholesky_vec of the element of the triangular matrix at coordinates (i,j)
+    function f(j) #return a slice (da x 1 x batchsize) containing the jth columns of the lower triangular cholesky decomposition of the covariance
+        tc_diag = softplus.(cholesky_vec[c2idx(j,j):c2idx(j,j),:,:])
+        tc_other = cholesky_vec[c2idx(j,j)+1:c2idx(j+1,j+1)-1,:,:]
+        zs = Flux.Zygote.ignore() do 
+            zeros(eltype(tc_diag), da - size(tc_other,1) - 1,1,batch_size)
+        end
+        [zs; tc_diag; tc_other]
+    end
+    return mapreduce(f, hcat, 1:da)
+end
+
+function trilcol(cholesky_vec::CuArray,da)
+    batch_size = size(cholesky_vec, 3)
+    c2idx(i,j) = ((2da-j)*(j-1))÷2+i #return the position in cholesky_vec of the element of the triangular matrix at coordinates (i,j)
+    function f(j) #return a slice (da x 1 x batchsize) containing the jth columns of the lower triangular cholesky decomposition of the covariance
+        tc_diag = softplus.(cholesky_vec[c2idx(j,j):c2idx(j,j),:,:])
+        tc_other = cholesky_vec[c2idx(j,j)+1:c2idx(j+1,j+1)-1,:,:]
+        zs = Flux.Zygote.ignore() do 
+            CUDA.zeros(eltype(tc_diag), da - size(tc_other,1) - 1,1,batch_size) 
```

Comment:
I think we can use `similar` to create an uninitialized array and then fill it with zero. So that we can avoid using `CUDA` explicitly here.

Reply:
Yes indeed, that was my first approach but backprop did not work because that's an array mutation. Only later I had to ignore CUDA.zeros as well and I did not think that I could have done that with the original implementation. SO yes, I'll make that change thanks!